plugins/filter_watermark/heap.c

```diff
@@ -0,0 +1,217 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2021 The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <fluent-bit/flb_time.h>
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_log.h>
+
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <stddef.h>
+
+#include "heap.h"
+#include "watermark.h"
+
+enum reheap_direction { DIR_UP, DIR_DOWN };
+
+static void reheap(struct c_heap_t *h, size_t root, enum reheap_direction dir) 
+{
+    size_t left;
+    size_t right;
+    size_t min;
+    int status;
+    void *tmp;
+
+    /* Calculate the positions of the children */
+    left = (2 * root) + 1;
+    if (left >= h->array_len)
+        left = 0;
+
+    right = (2 * root) + 2;
+    if (right >= h->array_len)
+        right = 0;
+
+    /* Check which one of the children is smaller. */
+    if ((left == 0) && (right == 0))
```

Comment:
Still some of your conditional statements don't follow the guidelines for brace usage (please find and fix the rest).